### PR TITLE
Check ORT Model Conversion and Preprocessing [Debuggability]

### PR DESCRIPTION
**Pull request by rayankrish**
Intention: Pass a single input to an ORT session and PyTorch model prior to any training steps to check that the outcome is the same, indicating that model conversion was successful. This is part of feature #779652, improve debuggability.

Model Cases and Expected Outcome:
1. `_enable_internal_postprocessing=False` and `_extra_postprocessing=None`: only the ort model and the torch model will be compared. Otherwise, the torch model is compared with an ORT model before postprocessing and a distinct ORT model from after postprocessing.
2. ORTTrainer is given an onnx model without any PyTorch model: no comparison occurs.
3. ORTTrainer is given a PyTorch model which includes a loss function. All outputs labeled in `output_desc` are compared between the torch model and the two onnx models respectively.
4. ORTTrainer is given a PyTorch model and an explicitly defined loss_fn: All outputs labeled in `output_desc`, except loss, are compared between the torch model and the two onnx models respectively.

Notes:
- This script has been tested on nvidia-bert (case 3) and a sample example (case 4). It has additionally been tested for a toy bert onnx model (case 2) and with the flags to prevent postprocessing (case 1)
- The PyTorch model is stored on the CPU (may result in better accuracy)
- The tolerance is currently hardcoded at 0.01. Each model output is compared in order with python's math.isclose() function.
- The script will give a warning (and continue training) if the comparison finds a difference greater than the tolerance.
- The currently comparison only occurs before the first training step to check convergence but can be easily extended to further training steps to measure divergence.

### CODE REVIEW

File: orttraining/orttraining/python/ort_trainer.py [hunk 1]
```diff
@@ -660,16 +663,38 @@ def __init__(self, model, loss_fn, model_desc, training_optimizer_name, map_opti
...
 
+
+    def _debug_init_session(self):
+        self.no_postprocess_onnx_model = None
+        if not self._enable_internal_postprocess or not self._extra_postprocess:
+            self.no_postprocess_onnx_model = self.onnx_model_
+                
+            self.no_postprocess_session, self.no_postprocess_train_io_binding, self.no_postprocess_eval_io_binding, self.no_postprocess_output_name, _, self.no_postprocess_output_types = \
+                create_ort_training_session_with_optimizer(
+                    self.onnx_model_, self.device_,
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:682]: `self.no_postprocess_onnx_model` instead of `self.onnx_model_`?

File: orttraining/orttraining/python/ort_trainer.py [hunk 2]
```diff
@@ -660,16 +663,38 @@ def __init__(self, model, loss_fn, model_desc, training_optimizer_name, map_opti
...
         # see prepare_input_and_fetches for more details.
         self.loss_scale_input_name = 'default_loss_scale_input_name'
 
         self._init_session()
 
+
+    def _debug_init_session(self):
+        self.no_postprocess_onnx_model = None
+        if not self._enable_internal_postprocess or not self._extra_postprocess:
+            self.no_postprocess_onnx_model = self.onnx_model_
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:678]: Isn't this a refernece instead of a copy? If self.onnx_model_ suffers post processing, wouldnt self.no_postprocess_onnx_model point to the post processed model too?

File: orttraining/orttraining/python/ort_trainer.py [hunk 3]
```diff
@@ -660,16 +663,38 @@ def __init__(self, model, loss_fn, model_desc, training_optimizer_name, map_opti
...
+                
+            self.no_postprocess_session, self.no_postprocess_train_io_binding, self.no_postprocess_eval_io_binding, self.no_postprocess_output_name, _, self.no_postprocess_output_types = \
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:680]: all the new class properties that are being created are private, so it is better to preprend an underscore on them (e.g. `self._no_postprocess_session`, etc

File: orttraining/orttraining/python/ort_trainer.py [hunk 4]
```diff
@@ -660,16 +663,38 @@ def __init__(self, model, loss_fn, model_desc, training_optimizer_name, map_opti
...
+                    deepspeed_zero_stage=self.deepspeed_zero_stage_,
+                    enable_grad_norm_clip=self.enable_grad_norm_clip_,
+                    frozen_weights=self.frozen_weights_, opset_version=self.opset_version_,
+                use_deterministic_compute=self._use_deterministic_compute)
+
     def _init_session(self):
         if self.onnx_model_ is None:
             return
 
+        if self._debug_conversion: self._debug_init_session()
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:696]: add a new line between the `if` and `then` clauses to mimic the same code style in this file

File: orttraining/orttraining/python/ort_trainer.py [hunk 5]
```diff
@@ -845,6 +870,56 @@ def _prepare_input_and_fetches(self, input_desc_with_, internal_learning_rate, i
 
         return input, fetches
 
+    def _debug_conversion_step(self, input, input_descs, output_desc):
+        # handle case if onnx model is provided instead of torch
+        if not self.torch_model_: return
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:875]: break in 2 lines

File: orttraining/orttraining/python/ort_trainer.py [hunk 6]
```diff
@@ -845,6 +870,56 @@ def _prepare_input_and_fetches(self, input_desc_with_, internal_learning_rate, i
...
+        cpu_data_list = []
+        from inspect import signature
+        # trailing inputs for ort optimizer are avoided by considering pytorch inputs directly
+        for i, (name, input_elem) in enumerate(zip(input_names, input)):
+            if i >= len(signature(self.torch_model_.forward).parameters): break
+            cpu_data_list.append(input_elem.to('cpu'))
+        cpu_data = tuple(cpu_data_list)
+
+        # torch inference step
+        with torch.no_grad(): torch_out = [self.torch_model_(*cpu_data)]
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:905]: break in 2 lines

File: orttraining/orttraining/python/ort_trainer.py [hunk 7]
```diff
@@ -845,6 +870,56 @@ def _prepare_input_and_fetches(self, input_desc_with_, internal_learning_rate, i
...
+        
+        # iterate through outputs, except loss when loss_fn is distinct from the model
+        if self.loss_fn_: torch_out.insert(0, torch.empty(1, 1))
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:908]: break in 2 lines

File: orttraining/orttraining/python/ort_trainer.py [hunk 8]
```diff
@@ -845,6 +870,56 @@ def _prepare_input_and_fetches(self, input_desc_with_, internal_learning_rate, i
 
         return input, fetches
 
+    def _debug_conversion_step(self, input, input_descs, output_desc):
+        # handle case if onnx model is provided instead of torch
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:874]: Maybe we can add an informational message here, something like 'Comparing inference output between PyTorch and ONNX models'

File: orttraining/orttraining/python/ort_trainer.py [hunk 9]
```diff
@@ -845,6 +870,56 @@ def _prepare_input_and_fetches(self, input_desc_with_, internal_learning_rate, i
...
+        with torch.no_grad(): torch_out = [self.torch_model_(*cpu_data)]
+        
+        # iterate through outputs, except loss when loss_fn is distinct from the model
+        if self.loss_fn_: torch_out.insert(0, torch.empty(1, 1))
+        for ((out_name, npp_out), pp_out, t_out) in zip(no_postprocess_session_results.items(), postprocess_session_results.values(), torch_out):
+            if out_name == 'loss' and self.loss_fn_: continue
+            no_postprocess_out = npp_out.cpu().numpy().flatten()
+            postprocess_out = pp_out.cpu().numpy().flatten()
+            t_out = t_out.cpu().numpy().flatten()
+            print("Comparing Values of {}".format(out_name))
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:914]: As the original model is pytorch and the final one is ONNX, I would change the order in which informatio is printed and let PyTorch data comes first: `print("PyTorch vs ONNX before/without post processing: ", end='')` (the next print has to be changed to reflect this inversion in order)

"No post-processing" also doesn make clear we are talking about the onnx model

File: orttraining/orttraining/python/ort_trainer.py [hunk 10]
```diff
@@ -845,6 +870,56 @@ def _prepare_input_and_fetches(self, input_desc_with_, internal_learning_rate, i
...
+            if self.no_postprocess_onnx_model:
+                print("No Post-Processing vs PyTorch: ", end='')
+                print(np.absolute(np.absolute(no_postprocess_out)-np.absolute(t_out)).mean())
+                print("Post-Processing vs PyTorch:    ", end='')
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:918]: Invert order and use a more descriptive message, like "PyTorch vs ONNX after/with post processing"

File: orttraining/orttraining/python/ort_trainer.py [hunk 11]
```diff
@@ -912,6 +987,8 @@ def train_step(self, *args, **kwargs):
         if not isinstance(input, (list, tuple)):
             input = (input,)
 
+        if self._debug_conversion and self.current_step is 1: self._debug_conversion_step(input, input_descs, output_desc)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:990]: what should happen after `_debug_conversion_step`? I believe we should end the script as the user will probably want to look the result and not continue the training.

Let's double check that requirement

File: orttraining/orttraining/python/ort_trainer.py [hunk 12]
```diff
@@ -845,6 +870,56 @@ def _prepare_input_and_fetches(self, input_desc_with_, internal_learning_rate, i
 
         return input, fetches
 
+    def _debug_conversion_step(self, input, input_descs, output_desc):
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:873]: What do you think of something like `_compare_inference_output_between_pytorch_onnx()` to describe what is being done. The public flag already makes it clear that it is a debug flag. This method is internal and can be more accurate in the name

File: orttraining/orttraining/python/ort_trainer.py (hunk 11 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/ort_trainer.py:990]: `self.current_step == 1` is recommended for `int` comparison

File: orttraining/orttraining/python/ort_trainer.py [hunk 13]
```diff
@@ -606,6 +606,9 @@ def __init__(self, model, loss_fn, model_desc, training_optimizer_name, map_opti
                Defaults to None
             use_invertible_layernorm_grad: use invertible layernorm grad
                Defaults to False
+            _debug_conversion: runs a single inference step on the postprocessed and not postprocessed model
+               to compare with the PyTorch model. A small difference in output indicates model converstion was successful.
```
[the remark this answers is not in the document]
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/ort_trainer.py:610]: Maybe it is important to inform this comparison is done on CPU too (CPU is more accurate than GPU - but slower too)

File: orttraining/orttraining/python/ort_trainer.py [hunk 14]
```diff
@@ -535,7 +535,7 @@ def __init__(self, model, loss_fn, model_desc, training_optimizer_name, map_opti
                  global_step=0, get_lr_this_step=None, loss_scaler=None, deepspeed_zero_stage=0,
                  enable_grad_norm_clip=True, frozen_weights=[], _opset_version=DEFAULT_OPSET_VERSION,
                  _enable_internal_postprocess=True, _extra_postprocess=None, _use_deterministic_compute=False,
-                 use_invertible_layernorm_grad=False):
+                 use_invertible_layernorm_grad=False, _debug_conversion=False):
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:538]: "conversion" is not very clear for user of what is being converted :)

`_debug_onnx_model_export` or `_debug_onnx_model_conversion`

File: orttraining/orttraining/python/ort_trainer.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/ort_trainer.py:610]: nit: typo "converstion"

It is mentioned post processed and non-post processed model, but it is not described as ONNX model

File: orttraining/orttraining/python/ort_trainer.py (hunk 2 above)
Comment (rayankrish) [reply orttraining/orttraining/python/ort_trainer.py:678]: I thought this might be copying the model because later when postprocessing, it does `self._onnx_model_ = postprocess.run_postprocess(self.onnx_model_)` . How would I explicitly copy by value?

File: orttraining/orttraining/python/ort_trainer.py [hunk 15]
```diff
@@ -902,23 +915,26 @@ def _debug_conversion_step(self, input, input_descs, output_desc):
...
+        with torch.no_grad():
+            torch_out = [self.torch_model_(*cpu_data)]
         
         # iterate through outputs, except loss when loss_fn is distinct from the model
-        if self.loss_fn_: torch_out.insert(0, torch.empty(1, 1))
+        if self.loss_fn_:
+            torch_out.insert(0, torch.empty(1, 1))
         for ((out_name, npp_out), pp_out, t_out) in zip(no_postprocess_session_results.items(), postprocess_session_results.values(), torch_out):
-            if out_name == 'loss' and self.loss_fn_: continue
+            if out_name == 'loss' and self.loss_fn_:
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:925]:
```suggestion
            if out_name == self.model_desc_.outputs[0].name and self.loss_fn_:
```

File: orttraining/orttraining/python/ort_trainer.py [hunk 16]
```diff
@@ -845,6 +879,63 @@ def _prepare_input_and_fetches(self, input_desc_with_, internal_learning_rate, i
...
         return input, fetches
 
+    def _compare_inference_output_pytorch_onnx(self, input, input_descs, output_desc):
+        # handle case if onnx model is provided instead of torch
+        if not self.torch_model_:
+            return
+
+        # set up training session for ort models (for inference)
+        no_postprocess_run_options = ort.RunOptions()
+        no_postprocess_run_options.only_execute_path_to_fetches = True
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:889]: We are not specifying `fetches=[]` on `train_step`/`eval_step`, so we probably dont need to specify `no_postprocess_run_options.only_execute_path_to_fetches=True` here

File: orttraining/orttraining/python/ort_trainer.py [hunk 17]
```diff
@@ -845,6 +879,63 @@ def _prepare_input_and_fetches(self, input_desc_with_, internal_learning_rate, i
...
+                                                                  no_postprocess_run_options)
+        else:
+            no_postprocess_session_results = postprocess_session_results # filler values not used
+
+        input_names = [elem.name_ for elem in input_descs]
+        cpu_data_list = []
+        from inspect import signature
+        # trailing inputs for ort optimizer are avoided by considering pytorch inputs directly
+        for i, (name, input_elem) in enumerate(zip(input_names, input)):
+            if i >= len(signature(self.torch_model_.forward).parameters): break
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:913]: break in 2 lines

File: orttraining/orttraining/python/ort_trainer.py (hunk 15 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/ort_trainer.py:925]: as the loss name is provided by the user and they are free to type whatever they want, we shouldn't assume 'loss' as the loss name.

What is known is that the first output is always loss, so we can use that

File: orttraining/orttraining/python/ort_trainer.py [hunk 18]
```diff
@@ -845,6 +879,63 @@ def _prepare_input_and_fetches(self, input_desc_with_, internal_learning_rate, i
...
+        # iterate through outputs, except loss when loss_fn is distinct from the model
+        if self.loss_fn_:
+            torch_out.insert(0, torch.empty(1, 1))
+        for ((out_name, npp_out), pp_out, t_out) in zip(no_postprocess_session_results.items(), postprocess_session_results.values(), torch_out):
+            if out_name == 'loss' and self.loss_fn_:
+                continue
+            no_postprocess_out = npp_out.cpu().numpy().flatten()
+            postprocess_out = pp_out.cpu().numpy().flatten()
+            t_out = t_out.cpu().numpy().flatten()
+            print("Comparing Values of {}".format(out_name))
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:930]: I was expecting that both `no_postprocess_out` and `postprocess_out` to be compared t_out and that the script would error out when it is higher than a certain tolerance (e.g. 0.01)

File: orttraining/orttraining/python/ort_trainer.py [hunk 19]
```diff
@@ -606,6 +606,12 @@ def __init__(self, model, loss_fn, model_desc, training_optimizer_name, map_opti
                Defaults to None
             use_invertible_layernorm_grad: use invertible layernorm grad
                Defaults to False
+            _debug_onnx_model_conversion: runs a single inference step on the postprocessed ONNX model  and not postprocessed ONNX model
+               to compare with the PyTorch model. A small difference in output indicates model conversion was successful.
+               If _enable_internal_postprocess is set to False and _extra_postprocess is None, the PyTorch model will only
+               be compared with the ONNX model before postprocessing. If the output after conversion is within a reasonable
+               threshold, training continues. If it isn't, an error is produced. The PyTorch model is stored and runs on CPU.
+               Defaults to False
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:614]: After our last meeting, I think we should also specify what is the tolerance to error out and exit the training script (0.01)

File: orttraining/orttraining/python/ort_trainer.py [hunk 20]
```diff
@@ -845,6 +883,91 @@ def _prepare_input_and_fetches(self, input_desc_with_, internal_learning_rate, i
...
+                # compare
+                import math
+                successful_conversion = True
+                for _torch, _ort in zip(t_out.tolist(), postprocess_out.tolist()):
+                    if not math.isclose(_torch, _ort, abs_tol=self._tolerance):
+                       warnings.warn("PyTorch model outputs differ from the ORT Model without Post Processing!")
+                       successful_conversion = False
+                       break
+                if successful_conversion:
+                    print("ORT model conversion successful")
```
Comment (rayankrish) [inline orttraining/orttraining/python/ort_trainer.py:970]: This makes the direct comparison without any calculations/normalizations. If this is reasonable, I can reduce duplicate code by making another method
